min/max-width/height, replaced elements

SimonSapin opened the pull request.

SimonSapin committed:
Move `clamp_*` functions to methods of `Length`

SimonSapin committed:
Replace `percent_resolved_*` functions with methods

SimonSapin committed:
Account for min/max-width in outer intrinsic sizing

SimonSapin committed:
Add min/max-width/height support for `inline-block`

(… and other non-replaced atomic inline-level boxes.)

SimonSapin committed:
Move replaced box used size computation to a method of `ReplacedContents`

SimonSapin committed:
Add sizing of inline replaced boxes

SimonSapin committed:
Don’t assume replaced elements have an intrinsic size

SimonSapin committed:
Fix min/max-content of replaced boxes

SimonSapin committed:
Remove inline/block_size from AbsolutelyPositionedFragment

SimonSapin committed:
Remove AbsoluteBoxOffsets’s type parameter

SimonSapin committed:
Rename ReplacedContent::used_size to used_size_as_if_inline_element

SimonSapin committed:
impl From<&'_ DefiniteContainingBlock> for ContainingBlock

SimonSapin committed:
Move two AbsoluteBoxOffsets fields into a Vec2

SimonSapin committed:
Implement replaced abspos

SimonSapin committed:
Delayed initialization over mutation

SimonSapin committed:
Conditionsals over closures

SimonSapin committed:
Move `solve_axis` function to module level

SimonSapin committed:
Struct with named fields over large tuple

highfive commented:
Heads up! This PR modifies the following files:
 * @jgraham: tests/wpt/metadata-layout-2020/css/CSS2/box-display/containing-block-008.xht.ini, tests/wpt/metadata-layout-2020/css/CSS2/box-display/display-change-001.xht.ini, tests/wpt/metadata-layout-2020/css/CSS2/box-display/containing-block-009.xht.ini, tests/wpt/metadata-layout-2020/css/CSS2/box-display/containing-block-028.xht.ini, tests/wpt/metadata-layout-2020/css/CSS2/box-display/containing-block-027.xht.ini and 1 more
 * @emilio: components/style/values/computed/length.rs, components/style/values/generics/length.rs

SimonSapin committed:
Fix a “Accessing content size that was not requested” panic

Percentage `width` are treated as `auto` for the purpose of
min/max-content computation, so they also need to be considered
when testing “wether width is auto”

nox commented on components/style/values/generics/length.rs:
```diff
@@ -207,6 +207,15 @@ impl<LengthPercentage> MaxSize<LengthPercentage> {
+
+    /// Convert
+    #[cfg(not(feature = "gecko"))]
+    pub fn to_option(self) -> Option<LengthPercentage> {
```
Nit: Just name that `length`, like `Result::ok`?

SimonSapin replied:
This is subtle but I mean this method not as "if this enum is the `LengthPercentage` variant, access its field" but rather "losslessly convert all the information in this enum". If/when we add support for e.g. `min-width: min-content` https://drafts.csswg.org/css-sizing/#min-size-properties this enum will have additional variants like it does today for Gecko, this method will need to be removed, and the callers should be modified to handle those variants separately.

nox commented on components/layout_2020/sizing.rs:
```diff
@@ -107,17 +107,37 @@ impl BoxContentSizes {
         // Percentages for 'width' are treated as 'auto'
         let inline_size = inline_size.map(|lp| lp.as_length());
         // The (inner) min/max-content are only used for 'auto'
         let mut outer = match inline_size.non_auto().flatten() {
```
Can't you use `and_then` to define `inline_size` and then remove the `flatten` call?

SimonSapin replied:
The call two lines above is `LengthPercentageOrAuto::map`, not `Option::map`. Maybe we should have `LengthPercentageOrAuto::and_then` that return `Auto` when its closure returns `Option::None`?

nox commented on components/layout_2020/dom_traversal.rs:
```diff
@@ -299,7 +298,7 @@ impl Drop for BoxSlot<'_> {
     fn is_element(self) -> bool;
     fn as_text(self) -> Option<String>;
-    fn as_image(self) -> Option<(Option<Arc<NetImage>>, Vec2<Length>)>;
+    fn as_image(self) -> Option<(Option<Arc<NetImage>>, Vec2<f64>)>;
```
Why the type change, and why not `CSSFloat`?

SimonSapin replied:
It’s not `Length` because it is measured in image pixels rather than CSS `px`.

I wanted to make it `u32` but it’s already divided by `script::dom::htmlimageelement::ImageRequest::current_pixel_density`. I’m not sure what that density means exactly or if diving there is correct.

It could be `CSSFloat`. It probably wouldn’t affect much since it’s not kept in memory for a long time.

nox commented on components/layout_2020/replaced.rs:
```diff
@@ -66,12 +93,21 @@ impl ReplacedContent {
+        let intrinsic_size = physical::Vec2 {
+            x: self.intrinsic_width,
+            y: self.intrinsic_height,
+        };
```
Why not store the intrinsic size as a `physical::Vec2<Option<_>>`?

SimonSapin replied:
Could be either. An earlier version of the comment on the field declaration talked about "all three components"

nox commented on components/layout_2020/replaced.rs:
```diff
@@ -82,25 +118,83 @@ impl ReplacedContent {
+        };
+        let clamp = |inline_size: Length, block_size: Length| Vec2 {
+            inline: inline_size.clamp_between_extremums(min_box_size.inline, max_box_size.inline),
+            block: block_size.clamp_between_extremums(min_box_size.block, max_box_size.block),
```
Wondering if we should have a meethod for that, don't we do that for non-replaced things elsewhere in the code?

SimonSapin replied:
I grepped through the code and didn’t find another instance of computing both at the same time.

In block layout for example, we compute the used width before recurring and the used height after.

nox commented on components/layout_2020/replaced.rs:
```diff
@@ -82,25 +118,83 @@ impl ReplacedContent {
         // https://drafts.csswg.org/css2/visudet.html#min-max-widths
         // https://drafts.csswg.org/css2/visudet.html#min-max-heights
-        let (inline_size, block_size) = match (box_size.inline, box_size.block) {
+        match (box_size.inline, box_size.block) {
```
Returning `Vec2<_>` values made the code more verbose, IMO, but it's not very important.

SimonSapin replied:
For some reason I thought this change helped having an early return, but it looks like it wasn’t necessary after all.

nox commented on components/layout_2020/replaced.rs:
```diff
@@ -82,25 +118,83 @@ impl ReplacedContent {
+                            // since it is available.
+                            block * i_over_b
+                        },
+                        // (Vec2 { inline: None, block: None }, Some(_)) => {
```
Commented code that is out-of-date, who would have known? :D Please remove.

SimonSapin replied:
What do you mean, out of date? The next line is a FIXME comment about a part of the spec that is not implemented yet, and this line has the match arm that we will need to use when we implement that part. I’ll shuffle this so it doesn’t look like comment-out code:

```rust
// FIXME
//
// “If 'height' and 'width' both have computed values of 'auto'
// and the element has an intrinsic ratio but no intrinsic height or width,
// […]”
//
// In this `match` expression this would be an additional arm here:
//
// ```
// (Vec2 { inline: None, block: None }, Some(_)) => {…}
// ```
//
// “[…] then the used value of 'width' is undefined in CSS 2.
// However, it is suggested that, if the containing block's width
// does not itself depend on the replaced element's width,
// then the used value of 'width' is calculated from the constraint
// equation used for block-level, non-replaced elements in normal flow.”

nox commented on components/layout_2020/replaced.rs:
```diff
@@ -82,25 +118,83 @@ impl ReplacedContent {
             },
             (LengthOrAuto::LengthPercentage(inline), LengthOrAuto::Auto) => {
-                clamp(inline, inline / intrinsic_ratio)
+                let block = if let Some(i_over_b) = intrinsic_ratio {
```
This name confuses me, please just name it `ratio`.

SimonSapin replied:
The problem these new names try to address is that when A divided by B is called a ratio, B divided by A is also called a ratio although they’re not interchangeable. Code that uses that ratio sometimes to multiple and sometimes to divide a length, having a variable name that indicates the “direction” of the ratio helps check if those are correct by doing pseudo-dimensional analysis.

We could declare that an aspect ratio is by convention always a width divided by a height. But this code work in flow-relative dimensions and accesses a ratio field in physical dimensions. I’ve added `inline_size_over_block_size_intrinsic_ratio` to make that conversion.

I’m open to other names, but I’d prefer if they indicate the “direction” of the ratio somehow. Or maybe we could provide that indication in doc-comments and code comments?

nox replied:
Fair enough, I don't mind the name.

nox commented on components/layout_2020/replaced.rs:
```diff
@@ -63,6 +64,41 @@ impl ReplacedContent {
+        &self,
+        style: &ComputedValues,
+    ) -> Option<CSSFloat> {
+        self.intrinsic_ratio.map(|width_over_height| {
```
Sorry, but again, I find the code more readable when that is just named `ratio`.

nox commented on components/layout_2020/positioned.rs:
```diff
@@ -158,7 +158,30 @@ impl<'a> AbsolutelyPositionedFragment<'a> {
+            Ok(replaced) => {
+                // https://drafts.csswg.org/css2/visudet.html#abs-replaced-width
+                // https://drafts.csswg.org/css2/visudet.html#abs-replaced-height
+                let u = replaced.used_size_as_if_inline_element(&containing_block.into(), style);
```
Not sur I like `u`.

nox commented on components/layout_2020/positioned.rs:
```diff
@@ -235,11 +235,12 @@ impl<'a> AbsolutelyPositionedFragment<'a> {
+                    let margin_end;
+                    let used_size;
+                    if let LengthOrAuto::LengthPercentage(s) = size {
+                        used_size = s;
```
Name `size` was just fine, why shorten it?

nox commented on components/layout_2020/positioned.rs:
```diff
@@ -208,7 +208,7 @@ impl<'a> AbsolutelyPositionedFragment<'a> {
             computed_margin_start: LengthOrAuto,
             computed_margin_end: LengthOrAuto,
-            solve_margins: impl FnOnce(Length) -> (Length, Length),
+            avoid_negative_margin_start: bool,
```
Please remove this change or introduce yet another new type for this boolean.

SimonSapin replied:
Even with exactly two call sites, that both have a comment to repeat the name of the parameter?

nox replied:
IIRC the collapsing margin stuff doesn't have more callers than that, and the comment can go if you introduce a type, but as I said on IRC, all my nits are optional.

nox reviewed (commented):
Lots of nits but otherwise nothing to say.

SimonSapin commented:
This conflicts with https://github.com/servo/servo/pull/25203, I’ll push a rebase + changes after that one lands.

SimonSapin commented:
@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 53ce714 has been approved by `nox`

<!-- @bors-servo r=nox 53ce7140059c07019489fc0eae522e9f76fcb08b -->

bors-servo commented:
:hourglass: Testing commit 53ce7140059c07019489fc0eae522e9f76fcb08b with merge 6cf618ae9fcc62a419ea8360520511cb5e083a94...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Uu85FoDqTf-FWJ8mgzwXfA)
Approved by: nox
Pushing 6cf618ae9fcc62a419ea8360520511cb5e083a94 to master...
